engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/IndirectArgumentSorterNode.java

```diff
@@ -41,10 +41,9 @@ private Object executeArguments(
       Object state,
       ThunkExecutorNode thunkExecutorNode) {
     for (int i = 0; i < mapping.getArgumentShouldExecute().length; i++) {
-      if (TypesGen.isThunk(arguments[i]) && mapping.getArgumentShouldExecute()[i]) {
+      if (mapping.getArgumentShouldExecute()[i]) {
```

Comment:
Having `mapping.getArgumentShouldExecute(i)` function would better encapsulate the underlaying `boolean[]` and would ensure it really cannot be changed from outside. Related:
- make the array fields `final` (in addition to have `@CompilationFinal(dimension=1)`)
- make `CallArgumentInfo` classes `final`
- `ArgumentMapping` constructor doesn't need to be `public` - make it package private at least to encourage/enforce usage of the builder

Reply:
I think you're right. Introduced the visibility and final restrictions. Not gonna touch the the array-to-method refactoring for now, because I can't in good conscience do it without running benchmarks (on the off-chance the compiler gets confused – I don't believe it would, but what if it does...) – and that would waste another hour of my time.